File: SS/Labs/Lab6/rdp_lab6.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

char input[200];
int pos = 0;
int error = 0;

void E();
void Eprime();
void T();
void Tprime();
void F();

void skip_space() {
    while (input[pos] == ' ' || input[pos] == '\t')
        pos++;
}

void E() {
    if (error) return;
    T();
    Eprime();
}

void Eprime() {
    if (error) return;
    skip_space();
    if (input[pos] == '+') {
        pos++;
        T();
        Eprime();
    }
}

void T() {
    if (error) return;
    F();
    Tprime();
}

void Tprime() {
    if (error) return;
    skip_space();
    if (input[pos] == '*') {
        pos++;
        F();
        Tprime();
    }
}

void F() {
    if (error) return;
    skip_space();

    if (input[pos] == '(') {
        pos++;
        E();
        skip_space();
        if (input[pos] == ')')
            pos++;
        else
            error = 1;
    }
    else if (isalpha((unsigned char)input[pos])) {
        while (isalnum((unsigned char)input[pos]))
            pos++;
    }
    else {
        error = 1;
    }
}

int main() {
    printf("Grammar:\n");
    printf("E  -> T E'\n");
    printf("E' -> + T E' | @\n");
    printf("T  -> F T'\n");
    printf("T' -> * F T' | @\n");
    printf("F  -> id | (E)\n\n");

    printf("Enter string: ");
    fgets(input, sizeof(input), stdin);

    if (strlen(input) > 0 && input[strlen(input) - 1] == '\n')
        input[strlen(input) - 1] = '\0';

    E();
    skip_space();

    if (!error && input[pos] == '\0')
        printf("Accepted\n");
    else
        printf("Rejected\n");

    return 0;
}

File: SS/Labs/PythonLabs/Lab6/rdp_lab6.py
# Lab 6 — Recursive Descent Parser (extended: supports identifiers only, not numbers)
# Grammar:
#   E  -> T E'
#   E' -> + T E' | @   (@ = epsilon)
#   T  -> F T'
#   T' -> * F T' | @
#   F  -> id | ( E )

class RDP:
    def __init__(self, text):
        self.inp = text.replace(' ', '').replace('\t', '')
        self.pos = 0
        self.error = False

    def peek(self):
        return self.inp[self.pos] if self.pos < len(self.inp) else '\0'

    def E(self):
        if self.error: return
        self.T()
        self.E_prime()

    def E_prime(self):
        if self.error: return
        if self.peek() == '+':
            self.pos += 1
            self.T()
            self.E_prime()

    def T(self):
        if self.error: return
        self.F()
        self.T_prime()

    def T_prime(self):
        if self.error: return
        if self.peek() == '*':
            self.pos += 1
            self.F()
            self.T_prime()

    def F(self):
        if self.error: return
        ch = self.peek()
        if ch == '(':
            self.pos += 1
            self.E()
            if self.peek() == ')':
                self.pos += 1
            else:
                self.error = True
        elif ch.isalpha():
            while self.pos < len(self.inp) and self.inp[self.pos].isalnum():
                self.pos += 1
        else:
            self.error = True

    def parse(self):
        self.E()
        return not self.error and self.pos == len(self.inp)

def main():
    print("Grammar:")
    print("E  -> T E'")
    print("E' -> + T E' | @")
    print("T  -> F T'")
    print("T' -> * F T' | @")
    print("F  -> id | (E)\n")

    s = input("Enter string: ").strip()
    parser = RDP(s)
    print("Accepted" if parser.parse() else "Rejected")

if __name__ == "__main__":
    main()
